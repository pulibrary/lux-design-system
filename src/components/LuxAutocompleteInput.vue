<template>
  <div class="lux-autocomplete">
    <label v-if="label" :class="{ 'lux-hidden': hideLabel }">{{ label }}</label>
    <div class="lux-autocomplete-input">
      <input
        id="displayInput"
        role="combobox"
        autocomplete="off"
        ref="autoComplete"
        type="text"
        @input="onChange"
        @click="onChange"
        @focus="onChange"
        v-model="search"
        @keydown.down.prevent="onArrowDown"
        @keydown.up.prevent="onArrowUp"
        @keydown.enter="onEnter"
        @keydown.esc="onEscape"
        @keydown.tab="onEscape"
        @blur="onEscape"
        :required="required"
        :aria-activedescendant="ariaActiveDescendant"
        :placeholder="placeholder"
      />
      <ul v-show="isOpen" class="lux-autocomplete-results">
        <li class="loading" v-if="isLoading">Loading results...</li>
        <li
          v-else
          v-for="(result, i) in results"
          :key="i"
          @mousedown="onClickResult(result)"
          class="lux-autocomplete-result"
          :class="{ 'is-active': i === arrowCounter }"
          :id="'lux-autocomplete-' + this.componentId + 'result-' + i"
        >
          {{ result }}
        </li>
      </ul>
      <input :name="name" :id="id" type="hidden" :value="inputValue" data-input-value />
    </div>
  </div>
</template>

<script>
import { useId } from "vue"

/**
 * InputAutocomplete is a cross between a text input and select input.
 * This component is used to offer users suggested values that
 * filter upon typing, while also allowing them to enter free-form text for the value.
 * The id and name supplied to this component are applied to a hidden input field, which
 * will contain the preferred value for submission based on the structure of the `items` prop.
 */
export default {
  name: "LuxAutocompleteInput",
  status: "prototype",
  release: "1.0.0",
  type: "Element",
  props: {
    /**
     * The available items in the autocomplete. This can be a simple array of strings
     * or an array of objects with an id and a label, if id is needed.
     */
    items: {
      type: Array,
      required: false,
      default() {
        return [""]
      },
    },
    /**
     * The placeholder value for the form input field.
     */
    placeholder: {
      type: String,
      default: null,
    },
    /**
     * The default value for the form input field.
     */
    defaultValue: {
      type: String,
      default: "",
    },
    /**
     * The label of the form input field.
     */
    label: {
      type: String,
      default: "",
    },
    /**
     * Visually hides the label of the form input field.
     */
    hideLabel: {
      type: Boolean,
      default: false,
    },
    /**
     * The id of the form input field.
     */
    id: {
      type: String,
      default: "",
    },
    /**
     * The name of the form input field.
     */
    name: {
      type: String,
      default: "",
    },
    /**
     * Is the data given by an outside ajax request?
     */
    isAsync: {
      type: Boolean,
      required: false,
      default: false,
    },
    /**
     * Whether the form input field is required or not.
     * `true, false`
     */
    required: {
      type: Boolean,
      default: false,
    },
    /**
     * Whether the input is focused or not
     * `true, false`
     */
    focused: {
      type: Boolean,
      default: false,
    },
  },
  emits: ["input", "selected"],
  data() {
    return {
      isOpen: false,
      results: [],
      search: "",
      inputValue: "",
      isLoading: false,
      arrowCounter: -1,
    }
  },
  methods: {
    onChange() {
      // Let's warn the parent that a change was made
      /**
       * Emitted on any input-like action that the user does: focusing the input, typing, etc.
       */
      this.$emit("input", this.search)

      // Is the data given by an outside ajax request?
      if (this.isAsync) {
        this.isLoading = true
      } else {
        // Data is sync, we can search our flat array
        this.filterResults()
        this.isOpen = true
      }
    },
    filterResults() {
      if (this.items.length && typeof this.items[0] === "object") {
        let preResults = this.items.filter(
          item => item.label.toLowerCase().indexOf(this.search.toLowerCase()) > -1
        )
        this.results = preResults.map(x => x.label)
      } else {
        this.results = this.items.filter(
          item => item.toLowerCase().indexOf(this.search.toLowerCase()) > -1
        )
      }
    },
    setResult(result, onSelectCallback = id => {}) {
      this.search = result
      this.inputValue = result
      this.isOpen = false

      if (this.items.length && typeof this.items[0] === "object") {
        // we need to search the input list for a matching label and return
        // an id if it's found
        let item = this.items.find(obj => {
          return obj.label === result
        })
        if (typeof item !== "undefined") {
          this.inputValue = item.id
          onSelectCallback(item.id)
        }
      }
    },
    onArrowDown() {
      if (this.arrowCounter < this.results.length) {
        this.arrowCounter = this.arrowCounter + 1
      }
    },
    onArrowUp() {
      if (this.arrowCounter > 0) {
        this.arrowCounter = this.arrowCounter - 1
      }
    },
    onEnter() {
      this.setResult(this.results[this.arrowCounter], this.emitSelectedId)
      this.isOpen = false
      this.arrowCounter = -1
    },

    onEscape() {
      this.setResult(this.search)
      this.isOpen = false
    },
    handleClickOutside(evt) {
      if (!this.$el.contains(evt.target)) {
        this.setResult(this.search)
        this.isOpen = false
        this.arrowCounter = -1
      }
    },
    onClickResult(result) {
      this.setResult(result, this.emitSelectedId)
    },
    emitSelectedId(id) {
      /**
       * Emitted when the user selects an "official" value from the list of supplied terms
       */
      this.$emit("selected", id)
    },
  },
  watch: {
    // Once the items content changes, it means the parent component
    // provided the needed data
    items: function (value, oldValue) {
      // we want to make sure we only do this when it's an async request
      if (this.isAsync) {
        this.results = value
        this.isOpen = true
        this.isLoading = false
      }
    },
  },
  computed: {
    ariaActiveDescendant() {
      if (this.arrowCounter < 0) {
        return null
      } else {
        return `lux-autocomplete-${this.componentId}result-${this.arrowCounter}`
      }
    },
  },
  created() {
    this.setResult(this.defaultValue)
  },
  mounted() {
    document.addEventListener("click", this.handleClickOutside)
    let vm = this

    vm.$nextTick(function () {
      if (vm.focused) {
        this.$refs.autoComplete.focus()
      }
    })
  },
  unmounted() {
    document.removeEventListener("click", this.handleClickOutside)
  },
  setup() {
    // A unique id that identifies this specific instance of the component, so that DOM ids are unique
    // even if you have many autocompletes on a screen
    const componentId = useId()
    return { componentId }
  },
}
</script>

<style lang="scss">
@use "/src/assets/styles/spacing.scss" as *;
@use "/src/assets/styles/variables.css" as *;
@use "/src/assets/styles/system.scss" as *;
// Design Tokens with local scope
$color-placeholder: tint($color-grayscale, 50%);

.lux-autocomplete {
  @include stack-space($space-small);
  font-weight: var(--font-weight-regular);
  font-family: var(--font-family-text);
  font-size: var(--font-size-base);
  line-height: var(--line-height-heading);
  position: relative;
  width: auto;

  &.lux-input-expand {
    width: 100%;
  }

  label {
    display: block;
    font-size: var(--font-size-base);
    color: tint($color-rich-black, 20%);
    @include stack-space($space-x-small);

    &.lux-hidden {
      @include visually-hidden;
    }
  }

  input {
    @include inset-squish-space($space-small);
    @include stack-space($space-small);
    appearance: none;
    -webkit-appearance: none;
    background: var(--color-white);
    border-radius: var(--border-radius-default);
    border: 0;
    color: set-text-color($color-rich-black, $color-white);
    font-family: var(--font-family-text);
    font-size: var(--font-size-base);
    font-weight: var(--font-weight-regular);
    line-height: var(--line-height-heading);
    margin: 0;
    transition: all 0.2s ease;
    width: 100%;

    &::-webkit-input-placeholder {
      -webkit-font-smoothing: antialiased;
      color: $color-placeholder;
    }
    &:-ms-input-placeholder {
      color: $color-placeholder;
    }
    &::-moz-placeholder {
      color: $color-placeholder;
      -moz-osx-font-smoothing: grayscale;
      opacity: 1;
    }

    &:hover,
    &[hover] {
      box-shadow: 0 1px 5px 0 color-mix(in srgb, var(--color-rich-black) 7%, transparent),
        0 0 0 1px tint($color-rich-black, 60%);
    }
    &:focus,
    &[focus] {
      transition: box-shadow 0.2s ease;
      box-shadow: inset 0 0 0 1px var(--color-bleu-de-france), 0 0 0 1px var(--color-bleu-de-france);
      outline: 0;
    }
  }
}

.lux-autocomplete-input {
  @include reset;
  background: var(--color-white);
  border-radius: var(--border-radius-default);
  box-shadow: inset 0 1px 0 0 color-mix(in srgb, var(--color-rich-black) 7%, transparent),
    0 0 0 1px tint($color-rich-black, 80%);
}

.lux-autocomplete-results {
  padding: 0;
  margin: 0;
  border: 1px solid var(--color-gray-10);
  height: 120px;
  overflow: auto;
  color: set-text-color($color-rich-black, $color-white);
  font-family: var(--font-family-text);
  font-size: var(--font-size-base);
  font-weight: var(--font-weight-regular);
  line-height: var(--line-height-heading);
  position: absolute;
  background: white;
  width: 100%;
  z-index: 999;
}

.lux-autocomplete-result {
  list-style: none;
  text-align: left;
  padding: 4px 2px;
  cursor: pointer;
}

.lux-autocomplete-result.is-active,
.lux-autocomplete-result:hover {
  background: var(--color-grayscale-light);
  color: var(--color-rich-black);
}
</style>

<docs>
  ```jsx
    <div>
    <lux-autocomplete-input label="Fruit" default-value="Banana" :items="[ 'Apple', 'Banana', 'Orange', 'Mango', 'Pear', 'Peach', 'Grape', 'Tangerine', 'Pineapple']" />
    <lux-autocomplete-input label="Fruit with IDs" default-value="Banana" :items="[ {id: 1, label: 'Apple'}, {id: 2, label: 'Banana'}, {id: 3, label: 'Mango'}, {id: 4, label: 'Pineapple'}]" />
    </div>
  ```
</docs>
