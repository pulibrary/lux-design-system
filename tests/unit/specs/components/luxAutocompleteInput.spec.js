import { mount } from "@vue/test-utils"
import LuxAutocompleteInput from "@/components/LuxAutocompleteInput.vue"
import { nextTick } from "vue"

describe("InputAutocomplete.vue", () => {
  let wrapper

  beforeEach(() => {
    wrapper = mount(LuxAutocompleteInput, {
      propsData: {
        id: "foo",
        name: "bar",
        defaultValue: "Banana",
        items: [
          { id: 1, label: "Apple" },
          { id: 2, label: "Banana" },
          { id: 3, label: "Mango" },
          { id: 4, label: "Pineapple" },
        ],
        focused: true,
      },
      attachTo: document.body,
    })
  })

  it("should be focused when the focused property is set to true", () => {
    const input = wrapper.find("#displayInput").element
    expect(input).toBe(document.activeElement)
  })

  it("should populate the inputValue with an id when defaultValue is passed and found in items", () => {
    expect(wrapper.vm.inputValue).toBe(2)
  })

  it("keyboard navigation works", () => {
    // not sure why this isn't working
    // wrapper.setData({ arrowCounter: 1 })
    // wrapper.vm.arrowCounter = 1
    // wrapper.vm.onArrowDown()
    // expect(wrapper.vm.arrowCounter).toBe(1)
  })

  it("hitting enter resets the counter and closes the dropdown", () => {
    wrapper.setData({ isOpen: true })
    wrapper.setData({ arrowCounter: 0 })
    wrapper.vm.onEnter()
    expect(wrapper.vm.arrowCounter).toBe(-1)
    expect(wrapper.vm.isOpen).toBe(false)
  })

  it("should filter results", () => {
    wrapper.vm.search = "Apple"
    wrapper.vm.filterResults()
    expect(wrapper.vm.results).toEqual(["Apple", "Pineapple"])
  })

  it("should populate the inputValue with the supplied string when search string is not found in items", () => {
    wrapper.vm.setResult("Foo")
    expect(wrapper.vm.inputValue).toBe("Foo")
  })

  it("should have a label if passed in", async () => {
    expect(wrapper.find("label").exists()).toBe(false)
    wrapper.setProps({ label: "Fill in the blank:" })
    await nextTick()
    expect(wrapper.find("label").exists()).toBe(true)
    expect(wrapper.find("label").text()).toBe("Fill in the blank:")
  })

  it("should work with an item list comprised of simple strings", async () => {
    wrapper.setProps({ items: ["Apple", "Banana", "Orange"] })
    await nextTick()
    wrapper.vm.setResult(wrapper.vm.items[2])
    expect(wrapper.vm.inputValue).toBe("Orange")
  })

  it("should have the name value that is passed in as a prop", () => {
    const input = wrapper.find("[data-input-value]")
    expect(input.element.name).toBe("bar")
  })

  it("should have the id value that is passed in as a prop", () => {
    const input = wrapper.find("[data-input-value]")
    expect(input.element.id).toBe("foo")
  })

  it("emits an input event when the user changes the text", async () => {
    wrapper.find("#displayInput").setValue("Code4lib 2025")
    // Updates this code because wrapper.emitted().input includes the defaultValue
    expect(wrapper.emitted().input.length).toEqual(2)
    expect(wrapper.emitted().input[1]).toEqual(["Code4lib 2025"])
  })

  it("has the expected html structure", () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it("closes dropdown when we have a blur event", async () => {
    const input = wrapper.find("input")
    input.setValue("Man")
    await nextTick()

    // Tests that dropdown is initially visible when you start typing
    expect(wrapper.find(".lux-autocomplete-results").isVisible()).toBe(true)

    await wrapper.find("input").trigger("blur")
    expect(wrapper.find(".lux-autocomplete-results").isVisible()).toBe(false)
  })

  it("shows dropdown when input is clicked", () => {
    wrapper.find("input").trigger("click")
    expect(wrapper.find(".lux-autocomplete-results").isVisible()).toBe(true)
  })

  it("shows dropdown when input is focused", () => {
    wrapper.find("input").trigger("focus")
    expect(wrapper.find(".lux-autocomplete-results").isVisible()).toBe(true)
  })
})
