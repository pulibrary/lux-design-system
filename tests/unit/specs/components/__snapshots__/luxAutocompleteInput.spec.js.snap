// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputAutocomplete.vue has the expected html structure 1`] = `
<div
  class="lux-autocomplete"
>
  <!--v-if-->
  <div
    class="lux-autocomplete-input"
  >
    <input
      autocomplete="off"
      id="displayInput"
      role="combobox"
      type="text"
    />
    <ul
      class="lux-autocomplete-results"
      style=""
    >
      
      <li
        class="lux-autocomplete-result"
        id="lux-autocomplete-v-0result-0"
      >
        Banana
      </li>
      
    </ul>
    <input
      data-input-value=""
      id="foo"
      name="bar"
      type="hidden"
      value="2"
    />
  </div>
</div>
`;
